### ISSUE TITLE
Shipping rates session storage prevents `woocommerce_package_rates` filter on AJAX calls

### ISSUE THREAD

**Issue by clytras** [woocommerce/woocommerce#19169]
## Prerequisites

- [x] I have searched for similar issues in both open and closed tickets and cannot find a duplicate
- [ ] The issue still exists against the latest `master` branch of WooCommerce on Github (this is **not** the same version as on WordPress.org!)
- [x] I have attempted to find the simplest possible steps to reproduce the issue
- [ ] I have included a failing test as a pull request (Optional)

## Steps to reproduce the issue

1. Create some Zones and add shipping methods
2. Add a filter hook for `woocommerce_package_rates` and remove rates from `$rates` argument based on payment methods (using `$woocommerce->session->chosen_payment_method`)
3. Use AJAX by triggering `update_checkout` (`$('body').trigger('update_checkout');`) to refresh the checkout form when changing payment method

## Expected/actual behavior

Expected: All shipping methods to update and refilter using `woocommerce_package_rates` filter hook

Actual: Filter `woocommerce_package_rates` isn't triggering and the filter does not applies the new selected payment methods

## Isolating the problem

- [ ] This bug happens with only WooCommerce plugin active
- [ ] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/)
- [x] I can reproduce this bug consistently using the steps above

I don't think this is a bug; maybe the filter hood isn't design for AJAX usage.
The whole problem is of the session caching shipping packages. At [`class-wc-shipping.php#L308`](https://github.com/woocommerce/woocommerce/blob/master/includes/class-wc-shipping.php#L308) we see that the `woocommerce_package_rates` result rates are stored to WC session, which is causing the problem and there is no revalidation and recalling of `woocommerce_package_rates` if the shipping packages have been stored in WC session but only if `woocommerce_shipping_debug_mode` option is set. My workaround is to create a `woocommerce_init` action hook and hardcode restore the session key for my shipping package without a hash so it revalidates and recalls the `woocommerce_package_rates` filter hook:

```
function th_wc_init() { 
    if(isset(WC()->session) && isset(WC()->session->set)) {
        $stored_rates = ['package_hash' => ''];
        WC()->session->set('shipping_for_package_0', $stored_rates);
    }
}; 
         
// add the action 
add_action( 'woocommerce_init', 'th_wc_init'); 
```
It would be nice to have an option to disable session storing for such filter generated outputs for cases when AJAX calls are made to revalidate session information such as shipping rates.

## WordPress Environment

<details>
```
### WordPress Environment ###

Home URL: https://feelcare.gr
Site URL: https://feelcare.gr
WC Version: 3.2.6
Log Directory Writable: ✔
WP Version: 4.9.4
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: –
WP Cron: ✔
Language: el

### Server Environment ###

Server Info: Apache
PHP Version: 7.0.27
PHP Post Max Size: 64 MB
PHP Time Limit: 60
PHP Max Input Vars: 5120
cURL Version: 7.19.7
NSS/3.27.1

SUHOSIN Installed: –
Max Upload Size: 64 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 3.2.6
WC Database Prefix: wp_
MaxMind GeoIP Database: ❌ Η βάση δεδομένων GeoIP της MaxMind δεν υπάρχει - η γεωτοποθεσία δε θα λειτουργεί. Μπορείτε να τη μεταφορτώσετε και να την εγκαταστήσετε χειροκίνητα από το http://dev.maxmind.com/geoip/legacy/geolite/ στην τοποθεσία: . Μεταβείτε κάτω στις "Μεταφορτώσεις" και μεταφορστώστε το αρχείο "Binary / gzip" δίπλα στο "GeoLite Country". Παρακαλούμε αποσυμπιέστε το GeoIP.dat.gz και ανεβάστε μόνο το αρχείο GeoIP.dat.
Συνολικό μέγεθος βάσης δεδομένων: 36.94MB
Μέγεθος βάσης δεδομένων: 27.36MB
Μέγεθος ευρετηρίου βάσης δεδομένων: 9.58MB
wp_woocommerce_sessions: Δεδομένα: 2.52MB + Ευρετήριο: 0.06MB
wp_woocommerce_api_keys: Δεδομένα: 0.02MB + Ευρετήριο: 0.03MB
wp_woocommerce_attribute_taxonomies: Δεδομένα: 0.02MB + Ευρετήριο: 0.02MB
wp_woocommerce_downloadable_product_permissions: Δεδομένα: 0.02MB + Ευρετήριο: 0.05MB
wp_woocommerce_order_items: Δεδομένα: 0.08MB + Ευρετήριο: 0.02MB
wp_woocommerce_order_itemmeta: Δεδομένα: 0.30MB + Ευρετήριο: 0.25MB
wp_woocommerce_tax_rates: Δεδομένα: 0.02MB + Ευρετήριο: 0.06MB
wp_woocommerce_tax_rate_locations: Δεδομένα: 0.02MB + Ευρετήριο: 0.05MB
wp_woocommerce_shipping_zones: Δεδομένα: 0.02MB + Ευρετήριο: 0.00MB
wp_woocommerce_shipping_zone_locations: Δεδομένα: 0.02MB + Ευρετήριο: 0.05MB
wp_woocommerce_shipping_zone_methods: Δεδομένα: 0.02MB + Ευρετήριο: 0.00MB
wp_woocommerce_payment_tokens: Δεδομένα: 0.02MB + Ευρετήριο: 0.02MB
wp_woocommerce_payment_tokenmeta: Δεδομένα: 0.02MB + Ευρετήριο: 0.03MB
wp_woocommerce_log: Δεδομένα: 0.02MB + Ευρετήριο: 0.02MB
wp_commentmeta: Δεδομένα: 0.17MB + Ευρετήριο: 0.03MB
wp_comments: Δεδομένα: 0.19MB + Ευρετήριο: 0.09MB
wp_cuteslider: Δεδομένα: 0.00MB + Ευρετήριο: 0.00MB
wp_links: Δεδομένα: 0.02MB + Ευρετήριο: 0.02MB
wp_options: Δεδομένα: 2.34MB + Ευρετήριο: 0.09MB
wp_piraeusbank_transactions: Δεδομένα: 0.02MB + Ευρετήριο: 0.00MB
wp_postmeta: Δεδομένα: 14.52MB + Ευρετήριο: 7.03MB
wp_posts: Δεδομένα: 4.52MB + Ευρετήριο: 0.83MB
wp_revslider_css: Δεδομένα: 0.13MB + Ευρετήριο: 0.00MB
wp_revslider_layer_animations: Δεδομένα: 0.02MB + Ευρετήριο: 0.00MB
wp_revslider_navigations: Δεδομένα: 0.02MB + Ευρετήριο: 0.00MB
wp_revslider_sliders: Δεδομένα: 0.05MB + Ευρετήριο: 0.00MB
wp_revslider_slides: Δεδομένα: 1.52MB + Ευρετήριο: 0.00MB
wp_revslider_static_slides: Δεδομένα: 0.02MB + Ευρετήριο: 0.00MB
wp_smush_dir_images: Δεδομένα: 0.02MB + Ευρετήριο: 0.03MB
wp_social_users: Δεδομένα: 0.02MB + Ευρετήριο: 0.02MB
wp_termmeta: Δεδομένα: 0.09MB + Ευρετήριο: 0.16MB
wp_terms: Δεδομένα: 0.08MB + Ευρετήριο: 0.13MB
wp_term_relationships: Δεδομένα: 0.13MB + Ευρετήριο: 0.09MB
wp_term_taxonomy: Δεδομένα: 0.11MB + Ευρετήριο: 0.13MB
wp_usermeta: Δεδομένα: 0.19MB + Ευρετήριο: 0.17MB
wp_users: Δεδομένα: 0.02MB + Ευρετήριο: 0.05MB
wp_yith_wcwl: Δεδομένα: 0.02MB + Ευρετήριο: 0.02MB
wp_yith_wcwl_lists: Δεδομένα: 0.02MB + Ευρετήριο: 0.03MB

### Post Type Counts ###

attachment: 1347
blocks: 8
custom-css-js: 1
customize_changeset: 76
featured_item: 8
nav_menu_item: 412
page: 243
portfolio: 12
post: 19
product: 678
product_variation: 1174
revision: 559
shop_coupon: 2
shop_order: 235
wafs: 1
wpcf7_contact_form: 5
zn_pb_templates: 2

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (14) ###

Contact Form 7: από τον/την Takayuki Miyoshi – 5.0
Simple Custom CSS and JS: από τον/την Diana Burduja – 3.14
Duplicate Post: από τον/την Enrico Battocchi – 3.2.1
Google Analytics Dashboard for WP (GADWP): από τον/την Alin Marcu – 5.2.2
Maintenance: από τον/την fruitfulcode – 3.6.1
Nextend Google Connect: από τον/την Roland Soos
Jamie Bainbridge – 1.6.2

Regenerate Thumbnails: από τον/την Alex Mills (Viper007Bond) – 3.0.2
WooCommerce Smart COD: από τον/την Stratos Vetsos – 999.4.9.1
Piraeus Bank WooCommerce Payment Gateway: από τον/την Enartia – 1.3.1 – Δεν έχει ελεχθεί στην τελευταία ενεργή έκδοση του WooCommerce
WooCommerce: από τον/την Automattic – 3.2.6 – Η έκδοση 3.3.3 είναι διαθέσιμη
WooSidebars: από τον/την WooThemes – 1.4.3
WP Mail SMTP: από τον/την WPForms – 1.2.5
WP Smush: από τον/την WPMU DEV – 2.7.8
Λίστα Επιθυμιών YITH WooCommerce: από τον/την YITHEMES – 2.2.1

### Settings ###

API Enabled: ✔
Force SSL: –
Currency: EUR (€)
Currency Position: right
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)


### WC Pages ###

Βάση καταστήματος: #139 - /shop/
Καλάθι: #141 - /cart/
Ταμείο: #143 - /checkout/
Ο λογαριασμός μου: #145 - /my-account/
Όροι και Προϋποθέσεις: #1534 - /%cf%8c%cf%81%ce%bf%ce%b9-%cf%80%cf%81%ce%bf%cf%8b%cf%80%ce%bf%ce%b8%ce%ad%cf%83%ce%b5%ce%b9%cf%82/

### Theme ###

Name: Flatsome Child
Version: 3.0
Author URL: 
Child Theme: ✔
Parent Theme Name: Flatsome
Parent Theme Version: 3.4.3
Parent Theme Author URL: http://www.uxthemes.com/
WooCommerce Support: ✔

### Templates ###

Overrides: flatsome/woocommerce/archive-product.php
flatsome/woocommerce/cart/cart-empty.php
flatsome/woocommerce/cart/cart.php
flatsome/woocommerce/checkout/form-checkout.php
flatsome/woocommerce/checkout/form-coupon.php
flatsome-child/woocommerce/checkout/payment-method.php
flatsome/woocommerce/checkout/thankyou.php
flatsome/woocommerce/content-product.php
flatsome/woocommerce/content-product_cat.php
flatsome/woocommerce/content-single-product.php
flatsome/woocommerce/global/breadcrumb.php
flatsome/woocommerce/global/quantity-input.php
flatsome/woocommerce/global/wrapper-end.php
flatsome/woocommerce/global/wrapper-start.php
flatsome/woocommerce/loop/loop-end.php
flatsome/woocommerce/loop/loop-start.php
flatsome/woocommerce/loop/orderby.php
flatsome/woocommerce/loop/pagination.php
flatsome/woocommerce/loop/result-count.php
flatsome/woocommerce/loop/sale-flash.php
flatsome/woocommerce/myaccount/form-login.php
flatsome/woocommerce/myaccount/navigation.php
flatsome/woocommerce/notices/error.php
flatsome/woocommerce/notices/notice.php
flatsome/woocommerce/notices/success.php
flatsome/woocommerce/product-searchform.php
flatsome/woocommerce/single-product/photoswipe.php
flatsome/woocommerce/single-product/price.php
flatsome/woocommerce/single-product/product-image.php
flatsome/woocommerce/single-product/product-thumbnails.php
flatsome/woocommerce/single-product/related.php
flatsome/woocommerce/single-product/review.php
flatsome/woocommerce/single-product/sale-flash.php
flatsome/woocommerce/single-product/share.php
flatsome/woocommerce/single-product/short-description.php
flatsome/woocommerce/single-product/tabs/tabs.php
flatsome/woocommerce/single-product/title.php
flatsome/woocommerce/single-product/up-sells.php
flatsome/woocommerce/single-product-reviews.php
flatsome/woocommerce/single-product.php
```
</details>

**Comment by mikejolley** [comment]
The filter will only run once, and from then on it will be cached. If you want this to be ran multiple times, you'd need to invalidate the cache either by:

- Adding something unique to the package so the hash mismatches
- Clearing the session value before this code runs
- Bumping the transient version `WC_Cache_Helper::get_transient_version( 'shipping' )`

Running the filter every time regardless of cache could make heavy logic run again e.g. getting shipping quotes, which would remove the benefits of having a cache i the first place. For that reason it would be better to invalidate it your side if you need different rates.

Thanks

**Comment by YoshiYo** [comment]
Hey @mikejolley 

How do you invalidate the package with calling the filter hook ?
I'm really new on this, and I'm not sure how to accomplish that. 

I really need a hot reload on this cache to feat with a specific algorithm depending of the current user connected with it's address.
